### PR TITLE
Fix infinite loop in Context

### PR DESCRIPTION
**Pull request by rictic**
Fixes https://github.com/lit/lit/issues/4158

### CODE REVIEW

File: packages/labs/context/src/test/context-provider_test.ts [hunk 1]
```diff
@@ -225,3 +225,49 @@ memorySuite('memory leak test', () => {
...
+    n1 = 'c';
+
+    render() {
+      return html`<p>${this.n1}</p>`;
+    }
+  }
+  customElements.define('r-consumer', Consumer);
+
+  class Container extends LitElement {
+    // uncomment for infinite loop:
```
Comment (AndrewJakubowicz) [inline packages/labs/context/src/test/context-provider_test.ts:257]: Remove comment?

File: packages/labs/context/src/test/context-provider_test.ts [hunk 2]
```diff
@@ -225,3 +225,49 @@ memorySuite('memory leak test', () => {
...
+    }
+  }
+  customElements.define('r-container', Container);
+
+  const provider = new Provider();
+  document.body.appendChild(provider);
+  await new Promise((r) => setTimeout(r, 100));
+  // This test passes if it doesn't get into an infinite loop of attempted
+  // repartenting of subscriptions.
+  // document.body.removeChild(provider);
```
Comment (AndrewJakubowicz) [inline packages/labs/context/src/test/context-provider_test.ts:272]: Commented out code

File: packages/labs/context/src/test/context-provider_test.ts [hunk 3]
```diff
@@ -225,3 +225,49 @@ memorySuite('memory leak test', () => {
     );
   });
 });
+
+test('regression test for https://github.com/lit/lit/issues/4158?', async () => {
```
Comment (AndrewJakubowicz) [inline packages/labs/context/src/test/context-provider_test.ts:229]: Nit: Can remove ending `?` in the url

File: packages/labs/context/src/lib/controllers/context-provider.ts [hunk 4]
```diff
@@ -115,7 +115,25 @@ export class ContextProvider<T extends Context<unknown, unknown>>
...
+      // than one ValueProvider for the same context on the same hostElement,
+      // they will each call the consumer, and since they will each have their
+      // own dispose function, a well behaved consumer will notice the change
+      // in dispose function and call their old one.
+      //
+      // This will cause the subscriptions to thrash, but worse, without this
+      // set check here, we can end up in an infinite loop, as we add and remove
+      // the same subscriptions onto the end of the map over and over.
+      if (seen.has(callback)) {
+        continue;
```
Comment (AndrewJakubowicz) [inline packages/labs/context/src/lib/controllers/context-provider.ts:134]: Per https://github.com/lit/lit/issues/4158#issuecomment-1702892539, do we support DEV_MODE warnings in context?

Would it be useful to log a dev mode warning that something might be bad?

File: packages/labs/context/src/test/context-provider_test.ts (hunk 2 above)
Comment (justinfagnani) [inline packages/labs/context/src/test/context-provider_test.ts:272]: remove or uncomment?

File: packages/labs/context/src/test/context-provider_test.ts [hunk 5]
```diff
@@ -225,3 +225,49 @@ memorySuite('memory leak test', () => {
...
+
+    render() {
+      return html`<r-consumer></r-consumer>`;
+    }
+  }
+  customElements.define('r-container', Container);
+
+  const provider = new Provider();
+  document.body.appendChild(provider);
+  await new Promise((r) => setTimeout(r, 100));
```
Comment (justinfagnani) [inline packages/labs/context/src/test/context-provider_test.ts:269]: what's the wait for? If you iloop, will it happen only after a render? I'd leave a comment on when the iloop will occur and see if you can tighten the wait here if possible, shouldn't a rAF or setTimeout(r, 0) do it?

**Review by justinfagnani — APPROVED** [review]
ncie!

File: packages/labs/context/src/lib/controllers/context-provider.ts (hunk 4 above)
Comment (AndrewJakubowicz) [reply packages/labs/context/src/lib/controllers/context-provider.ts:134]: Alternative: treat this case as an exceptional case, and throw in DEV_MODE.

**Review by augustjk — APPROVED** [review]
Yay

File: packages/labs/context/src/lib/controllers/context-provider.ts (hunk 4 above)
Comment (justinfagnani) [reply packages/labs/context/src/lib/controllers/context-provider.ts:134]: I think the consequences of this bug are so bad, and the fix so small, that we should keep it. Not everyone runs dev mode, unfortunately.

File: .changeset/clean-avocados-type.md [hunk 6]
```diff
@@ -0,0 +1,5 @@
+---
+'@lit-labs/context': patch
+---
+
+Fix an infinite loop that could result when declares multiple providers for a single context. This is an error, but we shouldn't go into an infinite loop over it.
```
[the remark this answers is not in the document]
Comment (rictic) [reply .changeset/clean-avocados-type.md]: +1, done!

File: packages/labs/context/src/test/context-provider_test.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (rictic) [reply packages/labs/context/src/test/context-provider_test.ts]: Good point! Done

Comment (rictic) [reply packages/labs/context/src/test/context-provider_test.ts]: Good point! Done

File: packages/labs/context/src/test/context-provider_test.ts (hunk 5 above)
Comment (rictic) [reply packages/labs/context/src/test/context-provider_test.ts]: Tightened it to a setTimeout of 0

File: packages/labs/context/src/test/context-provider_test.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (rictic) [reply packages/labs/context/src/test/context-provider_test.ts]: Yes, useful for debugging, but meant to uncomment it

Comment (rictic) [reply packages/labs/context/src/test/context-provider_test.ts]: Done

Comment (rictic) [reply packages/labs/context/src/test/context-provider_test.ts]: Done

File: packages/labs/context/src/test/context-provider_test.ts (hunk 3 above)
Comment (rictic) [reply packages/labs/context/src/test/context-provider_test.ts]: True!

**Pull request merged**